### PR TITLE
Clean up transaction key flow

### PR DESCRIPTION
**Pull request by rnicoll**

### CODE REVIEW

File: core/src/main/kotlin/net/corda/flows/TransactionKeyFlow.kt [hunk 1]
```diff
@@ -71,11 +71,10 @@ object TxKeyFlow {
         override fun call(): TxIdentities {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/flows/TransactionKeyFlow.kt]: Can you move `Provider` into CoreFlowHandlers.kt. This is only needed by the node and shouldn't be exposed to the client.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 2]
```diff
@@ -261,7 +260,7 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
             }
 
             // TODO: Investigate having class path scanning find this flow
-            registerInitiatedFlow(TxKeyFlow.Provider::class.java)
+            registerInitiatedFlow(TransactionKeyFlow.Provider::class.java)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: This is a core flow and as such shouldn't be using the annotations. Instead add the registration for this flow in `installCoreFlows`

**Comment by rnicoll** [comment]
I knew there was something I forgot in the rework - added a core flow handler now.

File: node/src/main/kotlin/net/corda/node/services/CoreFlowHandlers.kt [hunk 3]
```diff
@@ -122,3 +122,10 @@ class ContractUpgradeHandler(otherSide: Party) : AbstractStateReplacementFlow.Ac
         ContractUpgradeFlow.verify(oldStateAndRef.state.data, expectedTx.outRef<ContractState>(0).state.data, expectedTx.commands.single())
     }
 }
+
+class TransactionKeyHandler(val otherSide: Party) : FlowLogic<Unit>() {
+    @Suspendable
+    override fun call() {
+        subFlow(TransactionKeyFlow.Provider(otherSide))
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/CoreFlowHandlers.kt]: I mean move Provider here - there's no need to sub-flow it. Also remove its @InitiatedBy annotation.

File: core/src/main/kotlin/net/corda/flows/AbstractIdentityFlow.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package net.corda.flows
+
+import net.corda.core.flows.FlowLogic
+import net.corda.core.identity.Party
+
+abstract class AbstractIdentityFlow(val otherSide: Party, val revocationEnabled: Boolean): FlowLogic<TransactionIdentities>() {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/flows/AbstractIdentityFlow.kt]: space before the colon

File: core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package net.corda.flows
+
+import net.corda.core.identity.Party
+import net.corda.core.serialization.CordaSerializable
+
+@CordaSerializable
+data class TransactionIdentities(val identities: List<Pair<Party, AnonymisedIdentity>>) {
+    constructor(vararg identities: Pair<Party, AnonymisedIdentity>) : this(identities.toList())
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt]: Use `Arrays.asList` to avoid unnecessary copying

File: core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    init {
+        require(identities.size == identities.map { it.first }.toSet().size) { "Identities must be unique: ${identities.map { it.first }}" }
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt]: This requirement is basically what a `Map` gives you. Make `identities` a Map

File: node/src/main/kotlin/net/corda/node/services/CoreFlowHandlers.kt [hunk 7]
```diff
@@ -122,3 +123,21 @@ class ContractUpgradeHandler(otherSide: Party) : AbstractStateReplacementFlow.Ac
...
+    }
+
+    override val progressTracker: ProgressTracker = ProgressTracker(SENDING_KEY)
+
+    @Suspendable
+    override fun call(): TransactionIdentities {
+        val revocationEnabled = false
+        progressTracker.currentStep = SENDING_KEY
+        val legalIdentityAnonymous = serviceHub.keyManagementService.freshKeyAndCert(serviceHub.myInfo.legalIdentityAndCert, revocationEnabled)
+        val otherSideAnonymous = sendAndReceive<AnonymisedIdentity>(otherSide, legalIdentityAnonymous).unwrap { validateIdentity(it) }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/CoreFlowHandlers.kt]: Since TransactionKeyHandler is kicked off by a counter-party, it's return value can never used anyone. Therefore it doesn't make sense to receive the other side's identity. I would suggest you revert the type parameter of `AbstractIdentityFlow` to be `T` to allow this handler not to have to return anything.

File: core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt (hunk 6 above)
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt]: It's stored as a non-map type because it has to be serializable.

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/flows/TransactionIdentities.kt]: The only Map you can't serialise is HashMap. LinkedHashMap will work fine.

File: finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt [hunk 8]
```diff
@@ -47,11 +47,7 @@ object IssuerFlow {
             val issueRequest = IssuanceRequestState(amount, issueToParty, issueToPartyRef, anonymous)
             return sendAndReceive<AbstractCashFlow.Result>(issuerBankParty, issueRequest).unwrap { res ->
                 val tx = res.stx.tx
-                val recipient = if (anonymous) {
-                    res.identities.forParty(issueToParty).identity
-                } else {
-                    issueToParty
-                }
+                val recipient = res.identities?.get(issueToParty)?.identity ?: issueToParty
```
Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt]: Making `identities` nullable causes the need to have `?.` chains. Why not simply use an empty map?

File: finance/src/test/kotlin/net/corda/flows/CashPaymentFlowTests.kt [hunk 9]
```diff
@@ -57,7 +56,7 @@ class CashPaymentFlowTests {
         mockNet.runNetwork()
         val (paymentTx, identities) = future.getOrThrow()
         val states = paymentTx.tx.outputs.map { it.data }.filterIsInstance<Cash.State>()
-        val paymentState: Cash.State = states.single { it.owner == identities.forParty(payTo).identity }
+        val paymentState: Cash.State = states.single { it.owner == identities!!.get(payTo)!!.identity }
```
Comment (shamsasari) [inline finance/src/test/kotlin/net/corda/flows/CashPaymentFlowTests.kt]: Yes, definitely no the nullable `identities`

File: finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt (hunk 8 above)
Comment (rnicoll) [inline finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt]: We'd only take one ? off the chain, at the cost of having a map passed around that we aren't using. That doesn't seem better per se

File: finance/src/test/kotlin/net/corda/flows/CashPaymentFlowTests.kt (hunk 9 above)
Comment (rnicoll) [inline finance/src/test/kotlin/net/corda/flows/CashPaymentFlowTests.kt]: Again we'd only take one `!!` off, the previous version enforced that a party must be present in the data class, but in converting to a map it's now optional so we have to handle nulls anyway. Also this is a test, so if we get a null back unexpectedly it's intentionally going to fail.

File: core/src/main/kotlin/net/corda/flows/TransactionKeyFlow.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        progressTracker.currentStep = AWAITING_KEY
+        val legalIdentityAnonymous = serviceHub.keyManagementService.freshKeyAndCert(serviceHub.myInfo.legalIdentityAndCert, revocationEnabled)
+        serviceHub.identityService.registerAnonymousIdentity(legalIdentityAnonymous.identity, serviceHub.myInfo.legalIdentity, legalIdentityAnonymous.certPath)
+
+        // Special case that if we're both parties, a single identity is generated
+        val identities = LinkedHashMap<Party, AnonymisedIdentity>()
+        if (otherSide == serviceHub.myInfo.legalIdentity) {
+            identities.put(otherSide, legalIdentityAnonymous)
+        } else {
+            val otherSideAnonymous = receive<AnonymisedIdentity>(otherSide).unwrap { validateIdentity(otherSide, it) }
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/flows/TransactionKeyFlow.kt]: Can you try doing `sendAndReceive` here instead. I know the other side does it as well, but it should still work and would make this bit cleaner.

File: finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt (hunk 8 above)
Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt]: It's an extra complexity to the API that's not needed. How is a `null` map different to an empty one? Turns out it's the same thing. That's not a friendly API. The cost of an empty map is trivial compared to this.

Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/IssuerFlow.kt]: Alternatively, we can make `identities` just of type `Map` and then the serialisation layer do the HashMap checking.

**Comment by rnicoll** [comment]
Restructured how identities are passed around, and in particular realised the cash flows always want to pass back their counterparty only, so that's a lot cleaner now.

File: finance/src/main/kotlin/net/corda/flows/CashIssueFlow.kt [hunk 11]
```diff
@@ -41,15 +41,11 @@ class CashIssueFlow(val amount: Amount<Currency>,
...
-            subFlow(TxKeyFlow.Requester(recipient))
+            subFlow(TransactionKeyFlow(recipient))
         } else {
-            TxKeyFlow.TxIdentities(emptyList())
-        }
-        val anonymousRecipient = if (anonymous) {
-            txIdentities.forParty(recipient).identity
-        } else {
-            recipient
+            TransactionKeyFlow.EMPTY_IDENTITIES
```
Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/CashIssueFlow.kt]: This is a public mutable map that anyone can go in and add stuff to, so this definitely won't work. If you look at the construction of an empty LinkedHashMap you'll find that it doesn't create any internal objects, so it's actually cheap to create one.

However that's besides the point - why does TransactionKeyFlow need to return a Map anyway? Why not simply have a simple data class that holds the two identities?

Comment (rnicoll) [inline finance/src/main/kotlin/net/corda/flows/CashIssueFlow.kt]: It's for later extension when we do 3+ way flows and need to handle returning more complex sets of anonymised identities than simply "them" and "us".

Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/CashIssueFlow.kt]: Fair enough, but the point about the mutable singleton map still stands.

**Comment by rnicoll** [comment]
Switch the empty map to Collections.EMPTY_MAP, which works well for this I think. Let me know if this is okay to merge now please?

File: finance/src/main/kotlin/net/corda/flows/CashIssueFlow.kt [hunk 12]
```diff
@@ -41,15 +41,11 @@ class CashIssueFlow(val amount: Amount<Currency>,
...
-            subFlow(TxKeyFlow.Requester(recipient))
+            subFlow(TransactionKeyFlow(recipient))
         } else {
-            TxKeyFlow.TxIdentities(emptyList())
-        }
-        val anonymousRecipient = if (anonymous) {
-            txIdentities.forParty(recipient).identity
-        } else {
-            recipient
+            Collections.EMPTY_MAP as Map<Party, AnonymisedIdentity>
```
Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/CashIssueFlow.kt]: `emptyMap<Party, AnonymisedIdentity>()`

File: finance/src/main/kotlin/net/corda/flows/CashPaymentFlow.kt [hunk 13]
```diff
@@ -35,15 +35,11 @@ open class CashPaymentFlow(
...
-            subFlow(TxKeyFlow.Requester(recipient))
+            subFlow(TransactionKeyFlow(recipient))
         } else {
-            TxKeyFlow.TxIdentities(emptyList())
-        }
-        val anonymousRecipient = if (anonymous) {
-            txIdentities.forParty(recipient).identity
-        } else {
-            recipient
+            Collections.EMPTY_MAP as Map<Party, AnonymisedIdentity>
```
Comment (shamsasari) [inline finance/src/main/kotlin/net/corda/flows/CashPaymentFlow.kt]: ditto
